core/src/main/java/io/jenkins/pluginhealth/scoring/probes/AbstractDeprecatedAnnotationImportsCheckProbe.java

```diff
@@ -0,0 +1,83 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.CollectionUtils;
+
+/**
+ * The abstract probe checks for deprecated imports in a plugin.
+ */
+public abstract class AbstractDeprecatedAnnotationImportsCheckProbe extends Probe {
+    public static final int ORDER = SCMLinkValidationProbe.ORDER;
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractDeprecatedAnnotationImportsCheckProbe.class);
+
+    @Override
+    protected ProbeResult doApply(Plugin plugin, ProbeContext context) {
+        final Path scmRepository = context.getScmRepository();
+
+        /* The "maxDepth" is set to Integer.MAX_VALUE because a repository can have multiple modules with class files in it. We do not want to miss any ".java" file.  */
+        try (Stream<Path> paths = Files.find(scmRepository, Integer.MAX_VALUE, (path, $) ->
+            Files.isRegularFile(path) && path.getFileName().toString().endsWith(".java"))) {
+            Set<String> deprecatedPaths = paths.map(this::getFileNameWithDeprecatedImport)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toSet());
+
+            return deprecatedPaths.isEmpty()
+                ? ProbeResult.success(key(), String.format(getSuccessMessage() + " at %s plugin.", plugin.getName()))
+                : ProbeResult.failure(key(), String.format(getFailureMessage() + " at %s plugin for ", plugin.getName()) + deprecatedPaths.stream().sorted(Comparator.naturalOrder()).collect(Collectors.joining(", ")));
+        } catch (IOException ex) {
+            LOGGER.error("Could not browse the plugin folder during probe {}", key(), ex);
+            return ProbeResult.error(key(), String.format("Could not browse the plugin folder during {} probe.", plugin.getName()));
+        }
+    }
+
+    @Override
+    public String[] getProbeResultRequirement() {
+        return new String[]{SCMLinkValidationProbe.KEY};
+    }
+
+    /**
+     * Looks for all the deprecated imports in a path.
+     *
+     * @param path The file Path to check deprecated imports.
+     * @return name of the file that contains deprecated imports.
+     */
+    private String getFileNameWithDeprecatedImport(Path path) {
+        try (Stream<String> fileContent = Files.lines(path)) {
+            if (CollectionUtils.containsAny(fileContent.toList(), getListOfDeprecatedImports())) {
+                return path.getFileName().toString();
+            }
+        } catch (IOException ex) {
+            LOGGER.error("Could not browse the {} plugin folder during probe. {}", key(), ex);
+        }
+        return null;
+    }
+
+    /**
+     * @return a success message.
+     */
+    abstract String getSuccessMessage();
+
+    /**
+     * @return a failure message.
+     */
+    abstract String getFailureMessage();
+
+    /**
+     * @return a list of deprecated imports.
+     */
+    abstract List<String> getListOfDeprecatedImports();
```

Comment:
While we are at abstracting code, you can see that your abstract probe is actually not just allowing implementations to search for deprecated imports, but for any imports... Which opens the way for more probes to be implemented on top of that.

I believe if we keep this abstract probe, it should actually be called `AbstractJavaImportsCheckProbe` because that is what it is actually doing. It does not look just for annotation imports, or deprecated imports, but you can make it search for any imports.